Kernel/FileSystem/InodeWatcher.cpp

```diff
@@ -1,59 +1,67 @@
 /*
  * Copyright (c) 2018-2020, Andreas Kling <kling@serenityos.org>
+ * Copyright (c) 2021, the SerenityOS developers.
  *
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
 #include <AK/Memory.h>
 #include <Kernel/FileSystem/Inode.h>
 #include <Kernel/FileSystem/InodeWatcher.h>
+#include <Kernel/Process.h>
 
 namespace Kernel {
 
-NonnullRefPtr<InodeWatcher> InodeWatcher::create(Inode& inode)
+NonnullRefPtr<InodeWatcher> InodeWatcher::create()
 {
-    return adopt_ref(*new InodeWatcher(inode));
-}
-
-InodeWatcher::InodeWatcher(Inode& inode)
-    : m_inode(inode)
-{
-    inode.register_watcher({}, *this);
+    return adopt_ref(*new InodeWatcher);
 }
 
 InodeWatcher::~InodeWatcher()
 {
-    if (auto inode = m_inode.strong_ref())
-        inode->unregister_watcher({}, *this);
+    (void)close();
 }
 
 bool InodeWatcher::can_read(const FileDescription&, size_t) const
 {
-    return !m_queue.is_empty() || !m_inode;
-}
-
-bool InodeWatcher::can_write(const FileDescription&, size_t) const
-{
-    return true;
+    Locker locker(m_lock);
+    return !m_queue.is_empty();
 }
 
 KResultOr<size_t> InodeWatcher::read(FileDescription&, u64, UserOrKernelBuffer& buffer, size_t buffer_size)
 {
     Locker locker(m_lock);
-    VERIFY(!m_queue.is_empty() || !m_inode);
-
-    if (!m_inode)
-        return 0;
+    if (m_queue.is_empty())
+        // can_read will catch the blocking case.
+        return EAGAIN;
 
     auto event = m_queue.dequeue();
 
-    if (buffer_size < sizeof(InodeWatcherEvent))
-        return buffer_size;
+    // NOTE: unsigned so we can use it in memcpy
+    unsigned name_length = event.path.length() + 1;
```

Comment:
Please use `size_t` for buffer sizes.

Reply:
Should that also be a `size_t` on the `InodeWatcherEvent` itself? My main concern is having the variable size be the same size as the `name_length` in `InodeWatcherEvent`, and `name_length` can't possibly go beyond `NAME_MAX` unless something has horribly gone wrong.

Reply:
If the `name_length` can never go above `NAME_MAX`, then it can just as well be a `u16`, but here since we're doing arithmetic on a path length, let's use a `size_t`. :)